packages/actor-query-operation-group/lib/GroupsState.ts

```diff
@@ -44,56 +49,69 @@ export class GroupsState {
-      // Initialize state for all aggregators for new group
-      const aggregators: Record<string, AggregateEvaluator> = {};
-      for (const aggregate of this.pattern.aggregates) {
-        const key = termToString(aggregate.variable);
-        aggregators[key] = new AggregateEvaluator(aggregate, this.sparqleeConfig);
-        aggregators[key].put(bindings);
-      }
+    const groupTest: Promise<IGroup> | undefined = this.groupsInitializer.get(groupHash);
 
-      group = { aggregators, bindings: grouper };
-      this.groups.set(groupHash, group);
+    let res: Promise<any>;
+    if (!groupTest) {
+      // Initialize state for all aggregators for new group
+      const groupPromise = (async() => {
+        const aggregators: Record<string, AsyncAggregateEvaluator> = {};
+        await Promise.all(this.pattern.aggregates.map(async aggregate => {
+          const key = termToString(aggregate.variable);
+          aggregators[key] = new AsyncAggregateEvaluator(aggregate, this.sparqleeConfig);
+          await aggregators[key].put(bindings);
+        }));
 
-      if (this.distinctHashes) {
-        const bindingsHash = this.hashBindings(bindings);
-        this.distinctHashes.set(groupHash, new Set([ bindingsHash ]));
-      }
+        if (this.distinctHashes) {
+          const bindingsHash = this.hashBindings(bindings);
+          this.distinctHashes.set(groupHash, new Set([ bindingsHash ]));
+        }
+        return { aggregators, bindings: grouper };
+      })();
+      this.groupsInitializer.set(groupHash, groupPromise);
+      res = groupPromise.then(group => {
+        this.groups.set(groupHash, group);
+      });
     } else {
-      // Group already exists
-      // Update all the aggregators with the input binding
-      for (const aggregate of this.pattern.aggregates) {
-        // If distinct, check first wether we have inserted these values already
-        if (aggregate.distinct) {
-          const hash = this.hashBindings(bindings);
-          if (this.distinctHashes!.get(groupHash)!.has(hash)) {
-            continue;
-          } else {
+      const defGroupPromise = groupTest;
+      res = (async() => {
+        const group = await defGroupPromise;
+        await Promise.all(this.pattern.aggregates.map(async aggregate => {
+          // If distinct, check first whether we have inserted these values already
+          if (aggregate.distinct) {
+            const hash = this.hashBindings(bindings);
+            if (this.distinctHashes!.get(groupHash)!.has(hash)) {
+              return;
+            }
             this.distinctHashes!.get(groupHash)!.add(hash);
           }
-        }
 
-        const variable = termToString(aggregate.variable);
-        group.aggregators[variable].put(bindings);
-      }
+          const variable = termToString(aggregate.variable);
+          await group.aggregators[variable].put(bindings);
+        }));
+      })();
+    }
+    return res.then(() => {
+      this.subtractCounter();
```

Comment:
Instead of doing this via a separate `.then` call, could we just move this into the res promise a few lines above?

Reply:
I don't think this is possible because the promise above this is a list of promises. The subtract needs to happen after all those promises.

Reply:
You have an `await Promise.all` there. So if you move it after that promise resolves, it should be fine, right?

Reply:
oh you just mean before I put the promise into `res`? That would be possible. I would the just need to add it in both cases of the if.